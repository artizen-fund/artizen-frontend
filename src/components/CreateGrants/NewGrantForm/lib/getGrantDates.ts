import moment from 'moment-timezone'

export const getGrantDates = (startingDate: moment.Moment, length: number) => {
  /* Note on queueing grants!
     We want grants to start one second later (ex. 00:00:01) so that
     there is no overlap with a previous grant that ends at 00:00:00.
     
     sanitizeGrantDates('2022-12-20T00:00:00', 24 * 60)
     output: [
       2022-12-20T00:00:01,
       2022-12-21T00:00:00
     ]
  */

  const starting = startingDate.add(1, 's').format('YYYY-MM-DDTHH:mm:ss')
  const closing = startingDate.add(length, 'h').format('YYYY-MM-DDTHH:mm:ss')
  return [starting, closing]
}
