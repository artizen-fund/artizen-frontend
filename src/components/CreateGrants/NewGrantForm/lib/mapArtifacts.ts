import { Artifacts } from '@forms/createGrants'

export type MappedArtifact = {
  edition: 'community' | 'patron' | 'creator'
  artwork?: string
  video?: string
}

export const mapArtifactF = (artifactsData: Artifacts): Array<MappedArtifact> => {
  // TODO: This should be break down into smaller units
  // The ordering is specified here
  // https://docs.google.com/document/d/1Jwsmc3WxCK_Z-uRCjLNhTp8EbJGSF0CT7MNDbg0jmQk/edit
  // Creator, Community, Patron
  return [
    {
      edition: 'creator',
      artwork: artifactsData.artworkCreator,
      video: artifactsData.videoCreator,
    },
    {
      edition: 'community',
      artwork: artifactsData.artworkCommunity,
      video: artifactsData.videoCommunity,
    },
    {
      edition: 'patron',
      artwork: artifactsData.artworkPatron,
      video: artifactsData.videoPatron,
    },
  ]
}
