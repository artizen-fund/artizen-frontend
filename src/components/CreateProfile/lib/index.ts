import { useState } from 'react'
import { useMutation, useQuery, useReactiveVar } from '@apollo/client'
import { ErrorObject } from 'ajv'
import { useDebounce } from 'use-debounce'
import { ICheckForExistingArtizenHandleQuery, IUpdateUsersMutation, ICreateUsersMutation } from '@types'
import { loggedInUserVar, useCloudinary } from '@lib'
import { UPDATE_USER, CHECK_FOR_EXISTING_ARTIZENHANDLE, CREATE_USERS } from '@gql'
import { initialState, FormState } from '@forms/createProfile'

const useCreateProfile = () => {
  const { upload } = useCloudinary()

  const loggedInUser = useReactiveVar(loggedInUserVar)

  const [data, setData] = useState<FormState>(initialState)
  const [imageFile, setImageFile] = useState<File>()
  const [additionalErrors, setAdditionalErrors] = useState<Array<ErrorObject>>([])

  const [updateUser] = useMutation<IUpdateUsersMutation>(UPDATE_USER)
  const [createUser] = useMutation<ICreateUsersMutation>(CREATE_USERS)

  /* this checks for existing handles while the user types */
  const [newArtizenHandle] = useDebounce(data.artizenHandle, 500)
  useQuery<ICheckForExistingArtizenHandleQuery>(CHECK_FOR_EXISTING_ARTIZENHANDLE, {
    variables: {
      where: {
        _and: [{ artizenHandle: { _eq: newArtizenHandle?.toLowerCase() } }, { id: { _neq: loggedInUser?.id } }],
      },
    },
    onError: error => console.error('error ', error),
    fetchPolicy: 'no-cache',
    onCompleted: async ({ Users }) => {
      const errors: Array<ErrorObject> = []
      if (Users.length > 0) {
        errors.push({
          instancePath: '/artizenHandle',
          message: 'Handle is already in use',
          schemaPath: '#/properties/artizenHandle',
          keyword: '',
          params: {},
        })
      }
      setAdditionalErrors(errors)
    },
  })

  const createProfile = async () => {
    const profileImage = await uploadAvatar(imageFile)

    console.log('data  ', data)

    const createUserR = await createUser({
      variables: {
        objects: [
          {
            ...data,
            artizenHandle: data.artizenHandle?.toLowerCase(),
            profileImage,
          },
        ],
      },
    })
    await addUserToCourier()

    console.log('createUserR  ', createUserR)

    if (!createUserR.data?.insert_Users) {
      throw new Error('Error creating users in the admin form')
    }

    return createUserR.data?.insert_Users?.returning[0]
  }

  const updateProfile = async () => {
    if (!loggedInUser) {
      throw new Error('User session not found')
    }
    const profileImage = await uploadAvatar(imageFile)
    await updateUser({
      variables: {
        ...loggedInUser,
        ...data,
        artizenHandle: data.artizenHandle?.toLowerCase() || loggedInUser.artizenHandle,
        profileImage,
        claimed: true,
      },
    })
    await addUserToCourier()
  }

  const uploadAvatar = async (imageFile?: File) => {
    if (!imageFile) return undefined
    const cloudinaryResponse = await upload(imageFile)
    return cloudinaryResponse?.secure_url
  }

  const addUserToCourier = async () => {
    await fetch('/api/syncCourier', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        id: loggedInUser!.id,
        email: data.email,
        firstName: data.firstName,
        lastName: data.lastName,
      }),
    })
  }

  return { updateProfile, createProfile, additionalErrors, data, setData, setImageFile }
}

export default useCreateProfile
