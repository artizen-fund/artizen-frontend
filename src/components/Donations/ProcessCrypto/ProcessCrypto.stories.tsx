import ProcessCrypto from './'

export default {
  title: 'donations/ProcessCrypto',
  component: ProcessCrypto,
  argTypes: {},
}

export const ProcessCryptoComponent = (props: any) => <ProcessCrypto {...props} />
