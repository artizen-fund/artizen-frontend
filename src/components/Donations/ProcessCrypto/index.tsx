import { useState } from 'react'
import styled from 'styled-components'
import { IconStack, Icon } from '@components'
import { breakpoint, typography } from '@theme'

interface IProcessCrypto {
  setStage: (s: DonationStage) => void
}

type CryptoStage = 'swapping' | 'bridging' | 'building' | 'confirming' | 'complete'

const ProcessCrypto = ({ setStage }: IProcessCrypto) => {
  const [cryptoStage, setCryptoStage] = useState<CryptoStage>('swapping')

  return (
    <Wrapper>
      <Information>
        <div>
          <Title>It’s time to create your donation which requires a little extra magic</Title>
          <Subhead>
            Web3 takes longer to process transactions than traditional methods. You can keep this window open or close
            it without losing any progress – we&apos;ll send you an email when we’re ready.
          </Subhead>
        </div>

        <IconStack>
          <li onClick={() => setCryptoStage('bridging')}>
            <Icon outline={cryptoStage !== 'swapping'} glyph="swap" label="12% — Exchanging to USDC (est. 2m)" />
          </li>
          <li onClick={() => setCryptoStage('building')}>
            <Icon outline={cryptoStage !== 'bridging'} glyph="intersect" label="Bridging blockchains (est. 2m)" />
          </li>
          <li onClick={() => setCryptoStage('confirming')}>
            <Icon outline={cryptoStage !== 'building'} glyph="refresh" label="Building your donation (est. 10m)" />
          </li>
          <li onClick={() => setCryptoStage('complete')}>
            <Icon outline={cryptoStage !== 'confirming'} glyph="tick" label="Confirm your donation (est. 2m)" />
          </li>
          <li onClick={() => setStage('confirmation')}>
            <Icon outline={cryptoStage !== 'complete'} glyph="party" label="Donation Complete" />
          </li>
        </IconStack>
      </Information>

      <Distractions />
    </Wrapper>
  )
}

const Information = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  grid-area: copy;
`

const Title = styled.h1`
  ${typography.title.l2}
`

const Subhead = styled.h2`
  ${typography.body.l2}
`

const Wrapper = styled.div`
  display: grid;
  gap: 10px;
  grid-template-areas:
    'copy'
    'distractions';
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    gap: 12px;
    grid-template-columns: repeat(2, 1fr);
    grid-template-areas: 'copy distractions';
  }
  @media only screen and (min-width: ${breakpoint.desktop}px) {
    gap: 16px;
  }

  .vertical-layout,
  .vertical-layout-item {
    display: contents;
  }
`

const Distractions = styled.article`
  grid-area: distractions;
`

export default ProcessCrypto
