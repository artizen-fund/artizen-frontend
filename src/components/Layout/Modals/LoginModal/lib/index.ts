import { useConnect, useSignMessage, Connector } from 'wagmi'
import { useRouter } from 'next/router'
import { InjectedConnector } from 'wagmi/connectors/injected'
import { WalletConnectConnector } from '@wagmi/core/connectors/walletConnect'
import { assertInt, getWagmiClient } from '@lib'
import { signIn } from 'next-auth/react'

const useWalletConnect = () => {
  const { connectAsync } = useConnect()
  const { signMessageAsync } = useSignMessage()
  const { chains } = getWagmiClient()
  const router = useRouter()

  const connectWallet = async (connector: Connector) => {
    const chainId = assertInt(process.env.NEXT_PUBLIC_CHAIN_ID, 'NEXT_PUBLIC_CHAIN_ID')

    try {
      const { account: publicAddress, chain } = await connectAsync({
        connector,
        chainId,
      })

      const userData = { address: publicAddress, chain: chain.id, network: 'evm' }
      // todo: should "evm" be a constant? What does it mean?

      const response = await fetch('/api/auth/request-message', {
        method: 'POST',
        body: JSON.stringify(userData),
        headers: {
          'content-type': 'application/json',
        },
      })

      const { message } = await response.json()
      const signature = await signMessageAsync({ message })

      await signIn('credentials', { message, signature, redirect: false })

      // NOTE: this is necessary because of some Metamask logout bug that I don't understand.
      // Ruben: please document. -EJ
      router.reload()
    } catch (e) {
      console.error('error connecting user', e)
    }
  }

  const connectMetamask = () => connectWallet(new InjectedConnector({ chains }))

  const connectOtherWallet = () =>
    connectWallet(
      new WalletConnectConnector({
        chains,
        options: {
          qrcode: true,
        },
      }),
    )

  return { connectMetamask, connectOtherWallet }
}

export default useWalletConnect
