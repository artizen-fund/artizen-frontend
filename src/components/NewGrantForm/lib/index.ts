export * from './mapArtifactsForDB'
export * from './getGrantDates'
export { default as useSaveGrant } from './useSaveGrant'
export * from './validators'
