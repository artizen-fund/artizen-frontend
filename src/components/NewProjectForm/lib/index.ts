export * from './mapArtifactsForDB'
export * from './getGrantDates'
export { default as useSaveProject } from './useSaveProject'
export * from './validators'
