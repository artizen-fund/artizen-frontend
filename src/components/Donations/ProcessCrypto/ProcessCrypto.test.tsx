import renderer from 'react-test-renderer'
import { WagmiConfig } from 'wagmi'
import { getWagmiClient } from '../../../lib/wagmiClient'
import ProcessCrypto from './'

describe('ProcessCrypto', () => {
  // TODO: it's gonna take a bit of work to mock up a WagmiClient.
  // see https://github.com/tmm/testing-wagmi/blob/main/components/Connect.test.tsx

  // const setStage = (_: string) => {
  //   // do nothing
  // }
  // const setOrder = (_: { id: string }) => {
  //   // do nothing
  // }
  // const order = { id: 'derp' }
  // const donationMethod = 'ethereum'
  // const amount = 101
  //
  // it('renders correctly', () => {
  //   const { client, chains } = getWagmiClient()
  //
  //   const tree = renderer
  //     .create(
  //       <WagmiConfig client={client}>
  //         <ProcessCrypto {...{ setStage, chains, setOrder, order, donationMethod, amount }} />
  //       </WagmiConfig>,
  //     )
  //     .toJSON()
  //   expect(tree).toMatchSnapshot()
  // })

  it('does nothing', () => {
    expect(1).toEqual(1)
  })
})
