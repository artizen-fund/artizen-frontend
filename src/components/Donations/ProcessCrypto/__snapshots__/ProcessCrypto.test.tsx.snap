// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessCrypto renders correctly 1`] = `
<div
  className="sc-7f3ccf41-3 eimEzz"
>
  <div
    className="sc-7f3ccf41-0 becTpo"
  >
    <div>
      <h1
        className="sc-7f3ccf41-1 iVyQNF"
      >
        It’s time to create your donation which requires a little extra magic
      </h1>
      <h2
        className="sc-7f3ccf41-2 deGwoE"
      >
        Web3 takes longer to process transactions than traditional methods. You can keep this window open or close it without losing any progress – we'll send you an email when we’re ready.
      </h2>
    </div>
    <ol
      className="sc-d2bb1ef9-0 eXBKji"
    >
      <li
        onClick={[Function]}
      >
        <div
          className="sc-c089a03b-2 gfJPuX"
        >
          <div
            className="sc-c089a03b-0 eexWtz"
            color="white"
          >
            <div
              className="sc-149d3190-0 hIEUqx"
              color="white"
            />
          </div>
          <div
            className="sc-c089a03b-1 gmqtJg"
            color="night"
          >
            12% — Exchanging to USDC (est. 2m)
          </div>
        </div>
      </li>
      <li
        onClick={[Function]}
      >
        <div
          className="sc-c089a03b-2 gfJPuX"
        >
          <div
            className="sc-c089a03b-0 kNsEAy"
            color="night"
          >
            <div
              className="sc-149d3190-0 ikoVsw"
              color="night"
            />
          </div>
          <div
            className="sc-c089a03b-1 gmqtJg"
            color="night"
          >
            Bridging blockchains (est. 2m)
          </div>
        </div>
      </li>
      <li
        onClick={[Function]}
      >
        <div
          className="sc-c089a03b-2 gfJPuX"
        >
          <div
            className="sc-c089a03b-0 kNsEAy"
            color="night"
          >
            <div
              className="sc-149d3190-0 gbGGRA"
              color="night"
            />
          </div>
          <div
            className="sc-c089a03b-1 gmqtJg"
            color="night"
          >
            Building your donation (est. 10m)
          </div>
        </div>
      </li>
      <li
        onClick={[Function]}
      >
        <div
          className="sc-c089a03b-2 gfJPuX"
        >
          <div
            className="sc-c089a03b-0 kNsEAy"
            color="night"
          >
            <div
              className="sc-149d3190-0 fJmQeo"
              color="night"
            />
          </div>
          <div
            className="sc-c089a03b-1 gmqtJg"
            color="night"
          >
            Confirm your donation (est. 2m)
          </div>
        </div>
      </li>
      <li
        onClick={[Function]}
      >
        <div
          className="sc-c089a03b-2 gfJPuX"
        >
          <div
            className="sc-c089a03b-0 kNsEAy"
            color="night"
          >
            <div
              className="sc-149d3190-0 kTIffr"
              color="night"
            />
          </div>
          <div
            className="sc-c089a03b-1 gmqtJg"
            color="night"
          >
            Donation Complete
          </div>
        </div>
      </li>
    </ol>
  </div>
  <article
    className="sc-7f3ccf41-4 jlJDWf"
  />
</div>
`;
